sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -238,71 +238,77 @@ def build_task_spec_for_task(
-
-                additional_input_placeholder = placeholders.InputValuePlaceholder(
-                    additional_input_name)._to_string()
-                input_value = input_value.replace(channel.pattern,
-                                                  additional_input_placeholder)
-
-                if channel.task_name:
-                    # Value is produced by an upstream task.
-                    if channel.task_name in tasks_in_current_dag:
-                        # Dependent task within the same DAG.
-                        pipeline_task_spec.inputs.parameters[
-                            additional_input_name].task_output_parameter.producer_task = (
-                                utils.sanitize_task_name(channel.task_name))
-                        pipeline_task_spec.inputs.parameters[
-                            additional_input_name].task_output_parameter.output_parameter_key = (
-                                channel.name)
+        elif isinstance(input_value, (str, int, float, bool, dict, list)):
+            if isinstance(input_value, (str, dict, list)):
+                pipeline_channels = (
+                    pipeline_channel.extract_pipeline_channels_from_any(
+                        input_value))
+                for channel in pipeline_channels:
+                    # NOTE: case like this   p3 = print_and_return_str(s='Project = {}'.format(project))
+                    # triggers this code
+
+                    # value contains PipelineChannel placeholders which needs to be
+                    # replaced. And the input needs to be added to the task spec.
+
+                    # Form the name for the compiler injected input, and make sure it
+                    # doesn't collide with any existing input names.
+                    additional_input_name = (
+                        compiler_utils
+                        .additional_input_name_for_pipeline_channel(channel))
+
+                    # We don't expect collision to happen because we prefix the name
+                    # of additional input with 'pipelinechannel--'. But just in case
+                    # collision did happend, throw a RuntimeError so that we don't
+                    # get surprise at runtime.
+                    for existing_input_name, _ in task.inputs.items():
+                        if existing_input_name == additional_input_name:
+                            raise RuntimeError(
+                                f'Name collision between existing input name {existing_input_name} and compiler injected input name {additional_input_name}'
+                            )
+
+                    additional_input_placeholder = placeholders.InputValuePlaceholder(
+                        additional_input_name)._to_string()
+
+                    if isinstance(input_value, str):
+                        input_value = input_value.replace(
+                            channel.pattern, additional_input_placeholder)
```

Comment:
This would be covered by `recursive_replace_placeholders`, right?

Reply:
The string case is not covered by `recursive_replace_placeholders` as is, I would need to embed `string.replace()` logic in `recursive_replace_placeholders` if we wanted to get rid of this ifelse block.

I suppose question is if we want to expose this logic in `pipeline_spec_builder:build_task_spec_for_task` or `compiler_utils:recursive_repalce_placeholders`?

Reply:
I see. It's up to you whether you want to keep it as-is or remove the ifelse block. I don't have a strong preference.

I'm not sure how this is related to your question on exposing the logic.

Reply:
Alright, kept as it.